scrapy/core/http2/stream.py

```diff
@@ -59,26 +93,124 @@ def initiate_request(self):
 
             # TODO: Check if scheme can be "http" for HTTP/2 ?
             (":scheme", "https"),
-            (":path", url.path)
+            (":path", url.path),
+
+            # TODO: Make sure 'Content-Type' and 'Content-Encoding' headers
+            #  are sent for request having body
         ]
 
-        self._conn.send_headers(self.stream_id, http2_request_headers)
+        self._conn.send_headers(self.stream_id, http2_request_headers, end_stream=False)
+        self._write_to_transport()
+
+        self.send_data()
+
+    def send_data(self):
+        """Called immediately after the headers are sent. Here we send all the
+         data as part of the request.
+
+         If the content length is 0 initially then we end the stream immediately and
+         wait for response data.
+         """
+
+        # TODO:
+        #  1. Add test for sending very large data
+        #  2. Add test for small data
+        #  3. Both (1) and (2) should be tested for
+        #    3.1 Large number of request
+        #    3.2 Small number of requests
+
+        # Firstly, check what the flow control window is for current stream.
+        window_size = self._conn.local_flow_control_window(stream_id=self.stream_id)
+
+        # Next, check what the maximum frame size is.
+        max_frame_size = self._conn.max_outbound_frame_size
+
+        # We will send no more than the window size or the remaining file size
+        # of data in this call, whichever is smaller.
+        bytes_to_send = min(window_size, self.remaining_content_length)
+
+        # We now need to send a number of data frames.
+        while bytes_to_send > 0:
```

Comment:
:lipstick:  `bytes_to_send_size` would be less ambiguous, more consistent with similar variables in this function, and it would allow you to make this change without making the code more confusing:

```suggestion
        while bytes_to_send_size:
```